### PR TITLE
update wait documentation

### PR DESCRIPTION
**Pull request by titusfortner**
### Description
* Removed the static code examples and replaced with references code in the examples directory
* Added a section on hard coded sleeps; not technically a Selenium "feature," so maybe we don't want this?
* Removed Java-specific comments
* Removed the "Fluent Wait" section because that term only applies to Java (the example for customizing the explicit wait for Java uses the `FluentWait` class)
* Moved Expected Conditions out of the Wait page and into the Support Features page
* All explicit wait examples use lambdas not conditions classes
* Hopefully the explanations are more clear and focuses on the important bits

### Motivation and Context
* Code examples were not working
* Descriptions were repetitive and overly technical

### CODE REVIEW

**Review by harsha509 — DISMISSED** [review]
LGTM!

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 1]
```diff
@@ -1,495 +1,206 @@
...
-  });
-</script>
-```
-
-The WebDriver instructions might look innocent enough:
-
-{{< tabpane langEqualsHeader=true >}}
+Perhaps the most common challenge for browser automation is ensuring
+that the web application is in a state to execute a particular
+Selenium command as desired. the processes often end up in
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: 'The' processes -> new phrase should start with upper case

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 2]
```diff
@@ -1,495 +1,206 @@
...
+a _race condition_ where sometimes the browser gets into the right
+state first (things work as intended) and sometimes the Selenium code
+executes first (things do not work as intended). This is the
+primary cause of _flaky tests_.
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: Let's say it's 'one of the causes'. Others could be the environment, etc.

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 3]
```diff
@@ -1,495 +1,206 @@
...
+There are several synchronization strategies that can be used to properly wait for
+the application to be in the state you need it to be for the next Selenium command.
 
-Console.WriteLine(firstResult.Text);
-  {{< /tab >}}
-  {{< tab header="Ruby" >}}
-require 'selenium-webdriver'
-driver = Selenium::WebDriver.for :firefox
-wait = Selenium::WebDriver::Wait.new(:timeout => 10)
+## Hard-coded sleeps
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: Either remove this section or keep it but mark the fact that this is not a Selenium specific functionality.

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 4]
```diff
@@ -1,495 +1,206 @@
...
-Because the wait will swallow _no such element_ errors
-that are raised when the element is not found,
-the condition will retry until the element is found.
-Then it will take the return value, a _WebElement_,
-and pass it back through to our script.
+This is a global setting that applies to every element location call for the entire session.
+The default value is `0`, which means that if the element is not found, it will
+immediately return an error. If an implicit wait is set, the driver will wait for the 
+duration of the provided value before returning the error. Note that as soon as the 
+element is located, the driver will return the value and your code may continue, so a larger
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: "will return the element" i think? Or at least not the value cause i think the value refers to the timeout value?

Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: "continue executing"

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 5]
```diff
@@ -1,495 +1,206 @@
...
+implicit wait value won't necessarily increase the time of your session.
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: i would rephrase to say something like "increase the duration of the test execution"

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 6]
```diff
@@ -1,495 +1,206 @@
...
+added to a page or change visibility based on a click. For example, the box
+element takes a second to show up after clicking on the "adder" button, so
+even trying to locate that button will error without some form of synchronization: 
+
+{{< tabpane text=true langEqualsHeader=true >}}
   {{< tab header="Java" >}}
-driver.get("file:///race_condition.html");
-WebElement element = driver.findElement(By.tagName("p"));
-assertEquals(element.getText(), "Hello from JavaScript!");
+{{< gh-codeblock path="examples/java/src/test/java/dev/selenium/waits/WaitsTest.java#L23-L28" >}}
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: I would rewrite the test without the 'assert for exception', since the purpose of the test is not to make sure the assertion is thrown but for the element interaction to be possible. Below in a section there is a reference that says this code would not fail using a wait, but it's the other way around.

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 7]
```diff
@@ -1,495 +1,206 @@
...
-driver.implicitly_wait(10)
-driver.get("http://somedomain/url_that_delays_loading")
-my_dynamic_element = driver.find_element(By.ID, "myDynamicElement")
-  {{< /tab >}}
+## Explicit waits
+
+_Explicit waits_ are loops you add to your code that poll the application 
+for a specific condition to evaluate as true before it exits the loop and
+continues to the next command in your code. If the condition is not met before a designated timeout value, 
+the code will give a timeout error.
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: In Java it's an Exception. Not sure of other languages but if they all throw Exceptions i think that would be a better term. This is valid across this page (in multiple locations).

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 8]
```diff
@@ -1,495 +1,206 @@
...
+
+There are many ways for the application not to be in the desired state,
+so explicit waits are a great choice to specify exactly what is desired
+in each place they are needed.
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: I think this could be rephrased a bit.

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 9]
```diff
@@ -1,495 +1,206 @@
...
+the code will give a timeout error.
+
+There are many ways for the application not to be in the desired state,
+so explicit waits are a great choice to specify exactly what is desired
+in each place they are needed. 
+For example, an element must be both present in the DOM and 
+[displayed](({{< ref "elements/information/#is-displayed" >}})) on the page
+in order for Selenium to interact with it. 
+
+In this example, clicking the
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: Maybe "in the below example"

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 10]
```diff
@@ -1,495 +1,206 @@
...
+{{< badge-code >}}
   {{< /tab >}}
 {{< /tabpane >}}
 
-## FluentWait
+### Customization
 
-FluentWait instance defines the maximum amount of time to wait for a condition,
-as well as the frequency with which to check the condition.
+The Wait class can be created with various parameters that will change how the conditions are evaluated.
```
Comment (iamalittletester) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: instantiated? instead of created

**Comment by ollie-iterators** [comment]
If there are more areas that need their static code to be changed to references to the examples directory, I think there should be an issue added about the areas that need this to happen.

**Comment by titusfortner** [comment]
@ollie-iterators if we merge the PR I just linked we can tag the static code with `{{ badge-examples }}` to indicate that it needs to be moved over.

I'm planning to record some screen captures next week walking through each step of contributing to the documentation to better document what we're asking people to do to help.

**Comment by ollie-iterators** [comment]
> @ollie-iterators if we merge the PR I just linked we can tag the static code with `{{ badge-examples }}` to indicate that it needs to be moved over.
> 
> I'm planning to record some screen captures next week walking through each step of contributing to the documentation to better document what we're asking people to do to help.

Great.

File: website_and_docs/content/documentation/webdriver/waits.en.md (hunk 3 above)
Comment (titusfortner) [reply website_and_docs/content/documentation/webdriver/waits.en.md]: Good call, I updated the paragraph before and deleted this one. No need to have code examples.

File: website_and_docs/content/documentation/webdriver/waits.en.md (hunk 4 above)
Comment (titusfortner) [reply website_and_docs/content/documentation/webdriver/waits.en.md]: how about: "the driver will return the element reference"

File: website_and_docs/content/documentation/webdriver/waits.en.md (hunk 5 above)
Comment (titusfortner) [reply website_and_docs/content/documentation/webdriver/waits.en.md]: I'm trying to avoid use the word "session" instead of "test" to make it more "function-agnostic," but I think duration makes more sense than time.

File: website_and_docs/content/documentation/webdriver/waits.en.md (hunk 6 above)
Comment (titusfortner) [reply website_and_docs/content/documentation/webdriver/waits.en.md]: But you make a good point, I'm not writing a test suite showing that it fails when it is supposed to and passes when it is supposed to. Let me delete this code example and describe the scenario instead.

File: website_and_docs/content/documentation/webdriver/waits.en.md (hunk 7 above)
Comment (titusfortner) [reply website_and_docs/content/documentation/webdriver/waits.en.md]: Yeah, only Java & .NET make that distinction; plus it seems like the less technical term for describing what is happening.

File: website_and_docs/content/documentation/webdriver/waits.en.md (hunk 8 above)
Comment (titusfortner) [reply website_and_docs/content/documentation/webdriver/waits.en.md]: Good call:
> explicit waits are a great choice to specify the exact condition to wait for
in each place it is needed

**Comment by titusfortner** [comment]
Oof, this totally fell off my radar; I was going to do a quick update to get these changes out there, and of course... the structure is still confusing...

**Comment by titusfortner** [comment]
Ok, removed a bunch of repetitive things in explicit waits section, hopefully easier to read now.

**Review by iamalittletester — APPROVED** [review]

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 11]
```diff
@@ -1,495 +1,159 @@
...
+default value to wait for is `"complete"`) before the driver returns control to the code.
+The `readyState` only concerns itself with loading assets defined in the HTML, 
+but loaded JavaScript assets often result in changes to the site,
+and elements that need to be interacted with may not yet be on the page
+when the code is ready to execute the next Selenium command.
+
+Similarly, in a lot of single page applications, elements get dynamically
+added to a page or change visibility based on a click.
+An element must be both present and
+[displayed](({{< ref "elements/information/#is-displayed" >}})) on the page
```
Comment (harsha509) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: looks like an `( )` broke the link in here!

File: website_and_docs/content/documentation/webdriver/waits.en.md [hunk 12]
```diff
@@ -1,495 +1,159 @@
...
+can fail when it doesn't sleep long enough. Alternately, if the value is set too high
+and a sleep statement is added in every place it is needed, the duration of
+the session can become prohibitive.
+
+Selenium provides two different mechanisms for synchronization that are better.
+
+
+## Implicit waits
+Selenium has a built-in way to automatically wait for elements called an _implicit wait_.
+An implicit wait value can be set either with the [timeouts](({{< ref "drivers/options#timeouts" >}}))
```
Comment (harsha509) [inline website_and_docs/content/documentation/webdriver/waits.en.md]: here too, the link is broken due to extra `( )`

File: website_and_docs/content/documentation/webdriver/support_features/expected_conditions.en.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
+---
+title: "Waiting with Expected Conditions"
+linkTitle: "Expected Conditions"
+weight: 1
+description: >
+  These are classes used to describe what needs to be waited for.
+---
+
+Expected Conditions are used with [Explicit Waits](({{< ref "waits#explicit-waits" >}})).
```
Comment (harsha509) [inline website_and_docs/content/documentation/webdriver/support_features/expected_conditions.en.md]: extra `()` broke the link!

**Review by harsha509 — CHANGES_REQUESTED** [review]

**Comment by titusfortner** [comment]
odd there were a couple other places with double parens that seemed to work. Oh, I bet it was the combination double parens and the anchor reference at the end of the link...
Fixed all of them anyway

**Review by harsha509 — APPROVED** [review]
Thank you @titusfortner 🎉
